Improve error ouput (adding line number)

For example:

```
=> Starting doc-ui tests...

compare-elements... FAILED
[ERROR] assert didn't fail: for command `compare-elements-false: (".content>.right>p", ".content>.left", ["color"])`
```

It would be nice to have the line number.

--- a/src/commands.js
+++ b/src/commands.js
@@ -1425,6 +1425,7 @@ function parseContent(content, options) {
                     'wait': res['wait'],
                     'checkResult': res['checkResult'],
                     'original': line,
+                    'line_number': i + 1,
                 });
             }
         } else {

--- a/src/index.js
+++ b/src/index.js
@@ -191,24 +191,27 @@ async function runCommand(loaded, logs, options, browser) {
         const commands = loaded['commands'];
         for (let x = 0; x < commands.length; ++x) {
             let failed = false;
-            debug_log.append(`EXECUTING "${commands[x]['code']}"`);
+            const command = commands[x];
+            const line_number = command['line_number'];
+            debug_log.append(`EXECUTING (line ${line_number}) "${command['code']}"`);
             try {
-                await loadContent(commands[x]['code'])(page, extras).catch(err => {
+                await loadContent(command['code'])(page, extras).catch(err => {
                     failed = true;
                     const s_err = err.toString();
                     if (extras.expectedToFail !== true) {
-                        const original = commands[x]['original'];
-                        error_log = `[ERROR] ${s_err}: for command \`${original}\``;
+                        const original = command['original'];
+                        error_log = `[ERROR] (line ${line_number}) ${s_err}: for ` +
+                            `command \`${original}\``;
                     } else {
                         // it's an expected failure so no need to log it
-                        debug_log.append(`[EXPECTED FAILURE]: ${s_err}`);
+                        debug_log.append(`[EXPECTED FAILURE] (line ${line_number}): ${s_err}`);
                     }
                 });
             } catch (error) { // parsing error
-                error_log = 'output:\n' + error.message + '\n';
+                error_log = `(line ${line_number}) output:\n${error.message}\n`;
                 if (options.debug === true) {
-                    error_log += `command \`${commands[x]['original']}\` failed on ` +
-                        `\`${commands[x]['code']}\``;
+                    error_log += `command \`${command['original']}\` failed on ` +
+                        `\`${command['code']}\``;
                 }
             }
             debug_log.append('Done!');
@@ -216,13 +219,13 @@ async function runCommand(loaded, logs, options, browser) {
                 break;
             }
             if (failed === false
-                && commands[x]['checkResult'] === true
+                && command['checkResult'] === true
                 && extras.expectedToFail === true) {
-                error_log += `command \`${commands[x]['original']}\` was supposed to fail but ` +
-                    'succeeded';
+                error_log += `(line ${line_number}) command \`${command['original']}\` was ` +
+                    'supposed to fail but succeeded';
                 break;
             }
-            if (commands[x]['wait'] !== false) {
+            if (command['wait'] !== false) {
                 // We wait a bit between each command to be sure the browser can follow.
                 await page.waitFor(100);
             }

--- a/tests/test-js/api.js
+++ b/tests/test-js/api.js
@@ -752,10 +752,12 @@ function checkParseContent(x, func) {
             {
                 'code': 'await page.goto("file:///home");',
                 'original': 'goto: file:///home',
+                'line_number': 1,
             },
             {
                 'code': 'await arg.browser.overridePermissions(page.url(), arg.permissions);',
                 'original': 'goto: file:///home',
+                'line_number': 1,
             },
         ],
     });
@@ -772,14 +774,17 @@ function checkParseContent(x, func) {
                     'code': 'arg.expectedToFail = true;',
                     'wait': false,
                     'original': 'fail: true',
+                    'line_number': 1,
                 },
                 {
                     'code': 'await page.goto("file:///home");',
                     'original': 'goto: file:///home',
+                    'line_number': 2,
                 },
                 {
                     'code': 'await arg.browser.overridePermissions(page.url(), arg.permissions);',
                     'original': 'goto: file:///home',
+                    'line_number': 2,
                 },
             ],
         });

--- a/tests/test-js/exported_items.js
+++ b/tests/test-js/exported_items.js
@@ -1,3 +1,5 @@
+const fs = require('fs');
+const path = require('path');
 const process = require('process');
 
 const utils = require('../../src/utils.js');
@@ -399,11 +401,43 @@ async function checkOptions(x) {
     await x.assert(() => options20.validateFields());
 }
 
+// This test ensures that the outputs looks as expected.
+async function compareOutput(x, func) {
+    const filesToTest = [];
+    const testFolder = 'tests/ui-tests';
+    fs.readdirSync(testFolder).forEach(file => {
+        const curPath = path.join(testFolder, file);
+        if (fs.lstatSync(curPath).isDirectory() || !curPath.endsWith('.goml')) {
+            return;
+        }
+        filesToTest.push(curPath.toString());
+    });
+    for (let i = 0; i < filesToTest.length; ++i) {
+        const file = filesToTest[i];
+        const outputFile = file.replace('.goml', '.output');
+        let output;
+
+        try {
+            output = fs.readFileSync(outputFile, 'utf8');
+        } catch (_) {
+            x.addError(`Cannot open file \`${outputFile}\``);
+            continue;
+        }
+
+        const options = new Options();
+        options.parseArguments(['--variable', 'DOC_PATH', 'tests/html_files',
+            '--test-files', file]);
+
+        await x.assertTry(func, [options], [output, 1]);
+    }
+}
+
 const TO_CHECK = [
     {'name': 'Options', 'func': checkOptions},
     {'name': 'runTest', 'func': checkRunTest, 'toCall': wrapRunTest},
     {'name': 'runTestCode', 'func': checkRunTestCode, 'toCall': wrapRunTestCode},
     {'name': 'runTests', 'func': checkRunTests, 'toCall': wrapRunTests},
+    {'name': 'compareOutput', 'func': compareOutput, 'toCall': wrapRunTests},
 ];
 
 async function checkExportedItems(x = new Assert()) {

--- a/tests/test-js/utils.js
+++ b/tests/test-js/utils.js
@@ -48,10 +48,7 @@ class Assert {
     }
 
     assert(value1, value2, pos) {
-        if (this.testSuite.length > 0) {
-            this.testSuite[this.testSuite.length - 1].ranTests += 1;
-            this._incrField('totalRanTests', this.testSuite.length - 1);
-        }
+        this._addTest();
         if (typeof value2 !== 'undefined') {
             value1 = toJSON(value1);
             value2 = toJSON(value2);
@@ -73,10 +70,7 @@ class Assert {
                         break;
                     }
                 }
-                if (this.testSuite.length > 0) {
-                    this.testSuite[this.testSuite.length - 1].errors += 1;
-                    this._incrField('totalErrors', this.testSuite.length - 1);
-                }
+                this._incrError();
                 return;
             }
         } else if (!value1) {
@@ -84,11 +78,21 @@ class Assert {
                 pos = getStackInfo(new Error().stack);
             }
             print(`[${pos.file}:${pos.line}] failed: \`${value1}\` is evalued to false`);
-            if (this.testSuite.length > 0) {
-                this.testSuite[this.testSuite.length - 1].errors += 1;
-                this._incrField('totalErrors', this.testSuite.length - 1);
-            }
+            this._incrError();
             return;
+        }
+    }
+
+    addError(message) {
+        this._addTest();
+        print(`An error occurred: ${message}`);
+        this._incrError();
+    }
+
+    _addTest() {
+        if (this.testSuite.length > 0) {
+            this.testSuite[this.testSuite.length - 1].ranTests += 1;
+            this._incrField('totalRanTests', this.testSuite.length - 1);
         }
     }
 
@@ -99,6 +103,13 @@ class Assert {
             if (this.testSuite[pos][fieldName] !== undefined) {
                 this.testSuite[pos][fieldName] += 1;
             }
+        }
+    }
+
+    _incrError() {
+        if (this.testSuite.length > 0) {
+            this.testSuite[this.testSuite.length - 1].errors += 1;
+            this._incrField('totalErrors', this.testSuite.length - 1);
         }
     }
 

--- a/tests/ui-tests/line-number.goml
+++ b/tests/ui-tests/line-number.goml
@@ -1,0 +1,3 @@
+screenshot: false
+goto: file://|CURRENT_DIR|/|DOC_PATH|/elements.html
+compare-elements-false: (".content>.right>p", ".content>.left", ["color"])

--- a/tests/ui-tests/line-number.output
+++ b/tests/ui-tests/line-number.output
@@ -1,0 +1,7 @@
+=> Starting doc-ui tests...
+
+line-number... FAILED
+[ERROR] (line 3) assert didn't fail: for command `compare-elements-false: (".content>.right>p", ".content>.left", ["color"])`
+
+
+<= doc-ui tests done: 0 succeeded, 1 failed

--- a/tests/ui-tests/line-number2.goml
+++ b/tests/ui-tests/line-number2.goml
@@ -1,0 +1,4 @@
+screenshot: false
+goto: file://|CURRENT_DIR|/|DOC_PATH|/elements.html
+fail: true
+compare-elements: (".content>.right>p", ".content>.left", ["color"])

--- a/tests/ui-tests/line-number2.output
+++ b/tests/ui-tests/line-number2.output
@@ -1,0 +1,7 @@
+=> Starting doc-ui tests...
+
+line-number2... FAILED
+(line 4) command `compare-elements: (".content>.right>p", ".content>.left", ["color"])` was supposed to fail but succeeded
+
+
+<= doc-ui tests done: 0 succeeded, 1 failed

